### PR TITLE
More feature options, take two

### PR DESCRIPTION
**Pull request by dcbaker**
I'd previously opened some enhancements to feature in #10350, I've chosen to close that and open a new PR since this is a from scratch rewrite of the series (though with one of the same patches). This is specifically based on my experience of [porting mesa](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/20409) to use `feature.disable_auto_if` and `feature.require`.

I ran up against the core assumption of both `disable_auto_if` and `require` is that you're going to call `.allowed()` afterwards and convert to a boolean. I ran into a couple of cases where i *didn't* want to do that, but pass the feature to `dependency(required :)`. In such a case I don't want a boolean, I want a feature. As such I need to be able to force a dependency to enabled as well as disabled, but only the latter is currently possible. There is a workaround for this, but it's ugly

Consider the following case:
```meson
feat = get_option('feat') \
  .enable_auto_if(host_machine.system() == 'windows')  # If you turn this off on Windows you really know what you're doing \
  .disable_auto_if(host_machine.system() == 'haiku')  # Requires a nightly snapshot of haiku built with special options

dep = dependency('foo', required : feat)
```

You currently can't achieve this, easily, instead you' have to do something like:
```meson
feat = get_option('feat') \
  .disable_auto_if(host_machine.system() == 'haiku')  # Requires a nightly snapshot of haiku built with special options

dep = dependency('foo', required : feat.auto() and host_machine.system() == 'windows' ? true : feat)
```
That's ugly compared to:
```meson
feat = get_option('feat') \
  .disable_auto_if(host_machine.system() == 'haiku')  # Requires a nightly snapshot of haiku built with special options \
  .enable_auto_if(host_machine.system() == 'windows')  # 99% of people on Windows want this

dep = dependency('foo', required : feat)
```

The second case is similar, but related to `required`. I have some code in mesa like this (the actual code is in the commit message of the second commit):
```meson
_llvm = get_option('llvm')
dep_llvm = dependency('', required : false)
if not _llvm.disabled()
  dep_llvm = dependency(
    ...,
    required : with_x or with_y or with_z or _llvm.enabled()
  )
endif
if dep_llvm.found()
  ...
elif with_x
  error(...)
elif with_y
  error(...)
elif with_z
  error(...)
endif
```
In this case if `with_x` or `with_y` or `with_z` are true, then `_llvm` needs to be enabled. There is currently no way to achieve that. But with this series there is:
```meson
_llvm = get_option('llvm') \
  .enable_if(with_x, error_message : ...) \
  .enable_if(with_y, error_message : ...) \
  .enable_if(with_z, error_message : ...)
dep_llvm = dependency(
  ...,
  required : _llvm,
)
endif
if dep_llvm.found()
  ...
endif
```

Which is obviously a lot less code, easier to understand, and DRYer.  Not to mention that it would now be reasonable to add a call to `disable_auto_if()` to the end of that chain for cases where LLVM isn't actually used.

I have added as well a `disable_if` method. This is the thing I suspect that will be controversial, as `feature.disable_if(condition) == feature.require(not condition)`

I have a couple of justifications for this:

it is conceptually obvious when mixing `feature.enable_if` and `feature.disable_if`that these two methods do the opposite thing, especially when using `disable_auto_if` and `enable_auto_if`
consider:
```meson
f = get_option('feat') \
  .enable_if(foo) \
  .disable_if(foo2) \
  .enable_if_auto(bar)
```
vs
```meson
f = get_option('feat') \
  .enable_if(foo) \
  .require(not foo2) \
  .enable_if_auto(bar)
```
The first seems much more readable to me

`require` is hard (at least for me) to reason about. I expect require to be equivalent to `feature.enable_if(...)`, not `feature.disable_if(not ...)` These methods are less surprising to me (but I wrote them). This isn't as big of a deal if you plan to call `.allowed()` at the end, but if you want to keep them as features, I find it difficult to reason about.

@eli-schwartz, @bonzini: who were reviewing the previous PR

### CODE REVIEW

**Comment by dcbaker** [comment]
> For enable_auto_if I am totally missing the point. Auto features are intended to be silently disabled if absent, never silently enabled. So I am pretty sure it's conceptually wrong. This idea that auto will only ever be silently disabled may also clarify why require is named like that.

No, auto features are meant to be silently changed to a sane default based on the desire of the build system. They're a way for an end user to say "I don't know the right answer, please pick the best/most common answer". I would say they're like the Kernel's "If unsure say X here", but if KBuild just picked those for you.

For example, in mesa we have cases where we set things to "auto" in the meson_options.txt, but then set them to "disabled" on Windows and "enabled" on linux. An example of this would be the way we should handle glvnd, on windows and mac you *can't* use glvnd (windows and mac have their own system that solves the same problem, but it's part of the OS), but on Linux you should unless you really know what the heck you're doing. The average non-developer absolutely should enable glvnd, because basically all distros enable it, and mixing a mesa built glvnd support and one without requires a *lot* of care that the average "I just want this bug fix, or to be able to use my new card so I'm compiling mesa" end user probably doesn't know about (I actually ran into this helping someone this week, and they ended up with a broken system as a result).

```meson
_glvnd = get_option('glvnd') \
  .require(not host_machine.system() in ['darwin', 'windows'], error_message : 'GLVND not supported on Windows or MacOS') \
  .enable_auto_if(system_has_kms_drm)  # unless you know what you're doing you want this
dep_glvnd = dependency('glvnd', required : _glvnd)
```

There are two problems with your examples. I absolutely want to keep a feature, because I want the quick skip on disable of a feature. If I pass `dependency(.., required : FeatureOption[disabled])` meson will automatically return a not found dependency, but with the name set properly and some other niceties. Plus I don't have to do the
```meson
if disabled
  dep = null_dep
else
  dep = ...
```
Dance, because Meson does that internally. I'd like to (though It's a next step for that series) disable LLVM in the event that there are no consumers at all.

`with_gallium_radeonsi` is a boolean, not a feature. It is either "You must build this" or "you must _not_ build this" (the same is true for all of those options. If with_gallium_radeonsi is true, you *MUST* have LLVM. There is no other option than to error. In the trivial case that would look like this:
```
dependency('llvm', required : with_gallium_radeonsi)
```
```meson
llvm_err = _llvm.disabled() ? "LLVM is disabled" : "LLVM is not found"
# just use the feature here; or see below
dep_llvm = dependency('llvm', ..., required : _llvm)
# AFAIU this alone does not require LLVM? it could remain a boolean too
with_amd_vk = get_option('amd_vk') \
  .allowed()
# and if false aco_tests doesn't matter
with_aco_tests = with_amd_vk and get_option('aco_tests') \
  .require(llvm.found(), error_message: 'ACO tests require LLVM, but ' + llvm_err) \
  .allowed()
with_gallium_radeonsi = get_option('gallium_radeonsi')
  .require(llvm.found(), error_message: 'RadeonSI requires LLVM, but ' + llvm_err) \
  .allowed()

draw_with_llvm = with_aco_tests or with_gallium_radeonsi or with_gallium_opencl or with_clc
if not draw_with_llvm
  dep_llvm = null_dep
endif
```
This is problematic for two reasons because it makes the implementation detail drive the decision, rather than the decision drive the detail. LLVM is an implementation detail of radeonsi, llvmpipe, clover (opencl), etc. We want to turn LLVM support on because there is a driver that needs it. If I'm building just Intel (something I do a lot as an Intel developer), I don't want to even look for LLVM, because we don't use it), so converting from a feature to a boolean is bad, because then I have to do an if/else instead of just letting the feature do its job.

Second is that our list of gallium drivers is 21 elements long, and developers are adamant they want it as a list, not as 21 separate boolean options, so changing those into individual options is not on the table. (there's a separate list of vulkan drivers that's up to 10). 

There are other cases where it's optional, like `with_gallium_swrast, and there are cases where it's not needed at all, and we'd actually prefer not to find it or link with it. What I'd really like to end up with is:
```meson
_llvm = get_option('llvm') \
  .enable_if(with_x, error_message : ...) \
  .enable_if(with_y, error_message : ...) \
  .enable_if(with_z, error_message : ...) \
  .disable_auto_if(...) # we're not using a driver that can optionally use LLVM (such as with_gallium_swrast)
dep_llvm = dependency(
  ...,
  required : _llvm,
)
endif
if dep_llvm.found()
  ...
endif
```

**Comment by eli-schwartz** [comment]
Alternatively, one could simply pass `dependency('llvm', ...)` inline into each target that needs it, and it should only be looked up once due to caching. For cases where finding LLVM is an actual decision, use the feature option, but without any special frills.

**Comment by dcbaker** [comment]
There is *so* much logic around LLVM that that would be painful, it wouldn't just be
```meson
shared_library('foo', dependencies('llvm'))
```
it would end up being
```meson
shared_library(
  'foo',
  dependencies : dependency(
    'llvm',
    version : llvm_version,
    modules : llvm_modules
    optional_modules : llvm_opt_modules,
    required : <decision>,
    static : not shared_llvm,
    fallback : ...,
    include_type : 'system'))
```

**Comment by eli-schwartz** [comment]
> No, auto features are meant to be silently changed to a sane default based on the desire of the build system.

No, auto features are "meant" to do whatever is a sane default based on the desire of the end user as communicated by the build environment.

It's very possible mesa doesn't want to use feature options the way they are ostensibly intended to be used, but going from there to redefining what a feature option *is*, is exactly the issue I had in the previous version of this PR.

> I absolutely want to keep a feature, because I want the quick skip on disable of a feature.

This is why in the previous version of this PR, I suggested that it seems wrong to use feature options at all (because it's not an option!!! you don't want users to be allowed to define it on the command line) but instead you just want a way to create a custom feature object based on "the desires of the build system".

**Comment by eli-schwartz** [comment]
> And now every time someone changes the llvm dependency, they have to go check that ll of the calls to `dependency('llvm')` got updated. That's _horrifically_ worse than what we have now.

Hmm, indeed. I have in the past pondered a meson equivalent of functools.partial for exactly this sort of thing (and more). :)

**Comment by dcbaker** [comment]
> This is why in the previous version of this PR, I suggested that it seems wrong to use feature options at all (because it's not an option!!! you don't want users to be allowed to define it on the command line) but instead you just want a way to create a custom feature object based on "the desires of the build system".

It's an option that has interactions with other options. If you're building only Intel you don't need it. If you're building RadeonSI you do need it. If you're building swrast or r600 you *can* use it, but don't have to. There's also some fallback paths for old drivers that could use LLVM to make up for missing hardware features, but those drivers have all left the main tree so they don't matter so much...

**Comment by dcbaker** [comment]
> It's very possible mesa doesn't want to use feature options the way they are ostensibly intended to be used, but going from there to redefining what a feature option is, is exactly the issue I had in the previous version of this PR.

Then why did `feature.disable_if_auto` and `feature.require` land? I feel like we've added `and` and `or` to the language, and now people are complaining to me that I want to add `not` and `xor`.

**Comment by bonzini** [comment]
Ugh a long comment I was writing was eaten.

Anyhow, I think you _have_ made your point for enable_auto_if and enable_if. The crux of the issue is not the variables being booleans, but the options being lists.

I am just not sure of the naming for the latter. The issue is that enable_if and disable_if are not as symmetrical as it seems, because enabled->disabled makes sense but the latter does not. So the meaning of your enable_if is clear, the meaning of disable_if much less so. 

In other words, I would expect `required: opt.disable_if(cond)` to be similar to

```meson
  if cond
    foo = null_dep  # or disabler()
  else
    foo = dependency(..., required: opt)
```

and not to give an error if `cond` is true and `opt` is enabled.

However this is moot if disable_if is left out.

> every time someone changes the llvm dependency, they have to go check that ll of the calls to dependency('llvm') got updated

You could use kwargs, but I am not seriously suggesting it. :)

**Comment by bonzini** [comment]
> based on the desire of the end user as communicated by the build environment.

I think that's not different from what @dcbaker said: the end user communicates to the build system "pick a safe/sane default", the build system decides what the default should be based on the build environment.

It just so happens that Meson's only definition of "sane default" is "always yes if it exists". But there are ways to change that default from `meson.build`:

```meson
  x = null_dep
  if opt.enabled() or cond
    x = dependency (..., required: opt)
  endif

  y = dependency (...,
     required: opt.require(cond).allowed() ? true : opt)
```

and it makes sense to call these two algorithms `disable_auto_if` and `enable_auto_if` respectively.

**Comment by dcbaker** [comment]
> In other words, I would expect required: opt.disable_if(cond) to be similar to.

`feat.enable_if(cond, msg)` is equivalent to:
```meson
if con and feat.disabled()
  error(msg)
endif
feat = feat.enable_auto_if(true)
```

`feat.disable_if(cond, msg)` is equivalent to:
```meson
if con and feat.enabled()
  error(msg)
endif
feat = feat.disable_auto_if(true)
```
notably, (and this is where I'm personally tron on `disable_if`), this is also equivalent to:
```meson
feat.require(not cond, error_message : msg)
```

Which I'm torn on because I think that:
```meson
feat.require(not foo) \
      .enable_if(bar) \
      .require(other)
```
is harder to read than
```meson
feat.disable_iif(foo) \
     .enable_if(bar) \
     .disable_if(not other)
```
But that may just be personal taste.

**Comment by bonzini** [comment]
What is your use case for an enable_if/disable_if chain? The PR description doesn't have a real world case so I am unsure when you would encounter this idiom.

I understand that the implementation is symmetrical and the chain is more readable than with require, however I wouldn't have expected disable_if to have an error message parameter. enabled->disabled always makes sense unless you're verifying a requirement of the implementation of the option, for example it could be used for unnecessary dependencies or features. On the other hand enable_if should have the error message because disabled->enabled is invalid.

With a real world example however I could be persuaded easily...

**Comment by dcbaker** [comment]
> however I wouldn't have expected disable_if to have an error message parameter

I would consider any attempt to take something explicitly enabled and disable it as a potential error, ie:

```meson
with_thing = true
feat = get_option('feat') \
  .enable_if(with_thing) \
  .disable_if(host_machine.system() == 'darwin', error_message : 'feature is broken on MacOS')
dep_foo = dependency('foo', required : feat)
```
Which would be the same as if you chose to inverse the condition, and write `.disable_if()` then `enable_if()`. I think it's confusing if enable -> disable is valid, but disable -> enable is valid.

**Comment by bonzini** [comment]
Please give a *real world* example of enable_if, otherwise we're just talking past each other. What is "thing"? Is it a boolean option (or a list like in the mess driver case) which depends on feat? If so, wouldn't thing also be broken on Darwin?

**Comment by dcbaker** [comment]
Here is actualy code for Mesa I'm planning to propose once this lands:
```meson
_llvm = get_option('llvm') \
  .enable_if(with_amd_vk and with_aco_tests, error_message : 'ACO tests requires LLVM') \
  .enable_if(with_gallium_radeonsi, error_message : 'RadeonSI requires LLVM') \
  .enable_if(with_swrast_vk, error_message : 'Vulkan SWRAST requires LLVM') \
  .enable_if(with_gallium_opencl, error_message : 'The OpenCL Clover state trackers requires LLVM') \
  .enable_if(with_clc, error_message : 'CLC library requires LLVM') \
  .disable_auto_if(not (with_clc or with_gallium_opencl or with_swrast_vk or with_gallium_radeonsi or ...)) # etc
```
I expect (and the current Mesa meson.build does this) that setting `-Dllvm=disabled -Dgallium-drivers=radeonsi` (`with_gallium_radeonsi = get_option('gallium-drivers').contains('radeonsi')`), would result in an error, because llvm is explicitly disabled. 

Specifically though, I see cases of OS provided functionality as being the case for mixing .enable_if and disable_if:

```meson
use_ebpf = get_option('ebpf') \
    .disable_if(not host_machine.system() in ['linux', 'windows'], error_message : 'eBPF requires Linux or Windows') \
    .disable_if(cc.get_id() != 'gnu', error_message : 'Only GCC is currently supported for compiling eBPF programs') \
    .enable_if(get_option('run-really-fast'), error_message : 'option run-really-fast requires eBPF integration, but it cannot be enabled')
```

**Comment by bonzini** [comment]
Ok yeah, this is convincing (and it can even end with `.disable_auto_if(true)`). Please go ahead.

**Comment by dcbaker** [comment]
Unless someone objects I'm going to go ahead and merge this in the next 12 hours

**Comment by eli-schwartz** [comment]
There's a lot to unpack here and I seem to recall I was unhappy on a meta level about this because I suspected that what people really want is something like `make_feature_option(...)` and work from that. Trying to remember exactly what's up... I will try to take another look tonight.

**Comment by dcbaker** [comment]
I just can't seem to get anything beyond "turn the CI green" reviewed unless I complain loudly, so consider this me complaining loudly :)

I have a use case for this series, it's not a pragmatic thing. I *could* use a `make_feature_option()` in mesa for *other* cases, but in this case I really want the `.enable_if()` and `.disable_if()` methods, which I have enumerated ad nauseum above :)

**Comment by Akaricchi** [comment]
I drafted the original `make_feature_option()` proposal (#5206) back in the day, before we had `disable_auto_if()` and friends. The new methods turned out to have solved my use-cases for `make_feature_option()` much more elegantly. I could make some use of `enable_if` and `enable_auto_if` in Taisei as well. So consider this a thumbs up from me.

File: docs/markdown/snippets/feature_if_methods.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+## Add a FeatureOption.enable_if and .disable_if
+
+These are useful when features need to be constrained to pass to `dependency()`,
+as the behavior of an `auto` and `disabled` or `enabled` feature is markably
```
Comment (eli-schwartz) [inline docs/markdown/snippets/feature_if_methods.md:4]: markedly?

File: docs/markdown/snippets/feature_if_methods.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+different. consider the following case:
+
+```meson
+feat = get_option('feature').disable_auto_if(not foo)
```
Comment (eli-schwartz) [inline docs/markdown/snippets/feature_if_methods.md:8]: What does "feat" even mean? :eyes: Is that like an "opt"?

File: docs/yaml/objects/feature.yaml [hunk 3]
```diff
@@ -85,3 +85,63 @@ methods:
...
+  returns: feature
+  since: 1.1.0
+  description: |
+    Returns the object itself if the value is false; an error if the object is
+    `'disabled'` and the value is true; an enabled feature if the object
+    is `'auto'` or `'enabled'` and the value is true.
+
+  example: |
+    `enable_if` is useful to restrict the applicability of `'auto'` features,
+    particularly when passing them to [[dep]]:
```
Comment (eli-schwartz) [inline docs/yaml/objects/feature.yaml:99]: Passing them to `[[dependency]]` as in the function, not `[[dep]]` as in the returned object.

File: docs/yaml/objects/feature.yaml [hunk 4]
```diff
@@ -85,3 +85,63 @@ methods:
...
+    Returns the object itself if the value is false; an error if the object is
+    `'enabled'` and the value is true; a disabled feature if the object
+    is `'auto'` or `'disabled'` and the value is true.
+
+    This is equivalent to `feature_opt.require(not condition)`, but may make
+    code easier to reason about, especially when mixed with `enable_if`
+
+  example: |
+    `disable_if` is useful to restrict the applicability of `'auto'` features,
+    particularly when passing them to [[dep]]:
```
Comment (eli-schwartz) [inline docs/yaml/objects/feature.yaml:130]: Same.

File: docs/markdown/snippets/feature_if_methods.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
+## Add a FeatureOption.enable_if and .disable_if
+
+These are useful when features need to be constrained to pass to `dependency()`,
```
Comment (eli-schwartz) [inline docs/markdown/snippets/feature_if_methods.md:3]:
```suggestion
These are useful when features need to be constrained to pass to [[dependency]],
```

**Review by eli-schwartz — COMMENTED** [review]
Implementation review: some documentation nitpicks but other than that, it looks good.

In the commit message for enable_if:
> This is matters

Also:
> if the condition is true the returned value must be `enabled` or `disabled` respectively, auto will be removed

I was having a hard time parsing that sentence.

**Comment by eli-schwartz** [comment]
Thinking more about this, I'm still not convinced about the glvnd case, but the llvm case seems reasonable.

Anyway, I'm good with this enhancement being added once my documentation nits are tweaked.

File: docs/markdown/snippets/feature_if_methods.md (hunk 2 above)
Comment (dcbaker) [reply docs/markdown/snippets/feature_if_methods.md:8]: `feat` as a shortening of `feature` like `str` is a shortening of `string`? If it's not clear I can change it.

Comment (eli-schwartz) [reply docs/markdown/snippets/feature_if_methods.md:8]: It's not clear to me, but maybe I'm weird.

**Pull request merged**